src/open_petro_elastic/material/span_wagner/carbon_dioxide.py

```diff
@@ -0,0 +1,551 @@
+    argument documentation.
+    """
+    if dt == dd == 0:
+        _sum = sum(a0[3:] * log(1 - exp(-theta0 * tau)))
+        return log(delta) + a0[0] + a0[1] * tau + a0[2] * log(tau) + _sum
+    elif dt == 1 and dd == 0:
+        _sum = sum(a0[3:] * theta0 * ((1 - exp(-theta0 * tau)) ** -1 - 1))
+        return a0[1] + a0[2] / tau + _sum
+    elif dt == 0 and dd == 1:
+        return 1 / delta
+    elif dt == 2 and dd == 0:
+        _sum = sum(a0[3:] * theta0 ** 2 * exp(-theta0 * tau) * (1 - exp(-theta0 * tau)) ** -2)
+        return -a0[2] / tau ** 2 - _sum
+    elif dt == 0 and dd == 2:
+        return - 1 / delta ** 2
+    elif dt == 1 and dd == 1:
+        return 0
+    else:
+        raise NotImplementedError
+
+
+def phi_r(delta, tau, dd, dt):
+    """
+    Residual part of Helmholtz energy as defined by the equation in Table 32 of Span & Wagner [2]. See function phi for
+    argument documentation.
+    """
+    # tau == 1.0 or delta == 1.0 leads to numerically invalid results. The values are nudged to avoid nan output.
+    if tau == 1.0:
+        tau -= 1e-15
+    if delta == 1.0:
+        delta -= 1e-15
+
+    theta = A4*abs(delta - 1)**(1/beta4) - tau + 1
+    bigdelta = B4*abs(delta - 1)**(2*a4) + theta**2
+    bigphi = exp(-C4*(delta - 1)**2 - D4*(tau - 1)**2)
+    if dt == dd == 0:
+        s1 = sum(delta**d1*n1*tau**t1)
+        s2 = sum(delta**d2*n2*tau**t2*exp(-delta**c2))
+        s3 = sum(delta**d3*n3*tau**t3*exp(-alpha3*(delta - epsilon3)**2 - beta3*(-gamma3 + tau)**2))
+        s4 = sum(bigdelta**b4*bigphi*delta*n4)
+        return s1 + s2 + s3 + s4
+    elif dt == 1 and dd == 0:
+        dt_theta = -1
+        dt_bigdelta = 2*dt_theta*theta
+        dt_bigphi = -D4*(2*tau - 2)*exp(-C4*(delta - 1)**2 - D4*(tau - 1)**2)
+        dt_s1 = sum(delta**d1*n1*t1*tau**(t1 - 1))
+        dt_s2 = sum(delta**d2*n2*t2*tau**(t2 - 1)*exp(-delta**c2))
+        dt_s3 = sum(-beta3*delta**d3*n3*tau**t3*(-2*gamma3 + 2*tau)*exp(-alpha3*(delta - epsilon3)**2 - beta3*(
+                -gamma3 + tau)**2) + delta**d3*n3*t3*tau**(t3 - 1)*exp(-alpha3*(delta - epsilon3)**2 - beta3*(-gamma3 +
+                                                                                                              tau)**2))
+        dt_s4 = sum(b4*bigdelta**(b4 - 1)*bigphi*delta*dt_bigdelta*n4 + bigdelta**b4*delta*dt_bigphi*n4)
+        return dt_s1 + dt_s2 + dt_s3 + dt_s4
+    elif dt == 0 and dd == 1:
+        dd_theta = A4 * abs(delta - 1) ** (-1 + 1 / beta4) * sign(delta - 1) / beta4
+        dd_bigdelta = 2 * B4 * a4 * abs(delta - 1) ** (2 * a4 - 1) * sign(delta - 1) + 2 * dd_theta * theta
+        dd_bigphi = -C4 * (2 * delta - 2) * exp(-C4 * (delta - 1) ** 2 - D4 * (tau - 1) ** 2)
+        dd_s1 = sum(d1 * delta ** (d1 - 1) * n1 * tau ** t1)
+        dd_s2 = sum(-c2 * delta ** (c2 + d2 - 1) * n2 * tau ** t2 * exp(-delta ** c2) + d2 * delta ** (
+                    d2 - 1) * n2 * tau ** t2 * exp(-delta ** c2))
```

Comment:
Likely some part (or all, or some super-part :P )  of this needs to be made into a look up table, probably something like the following could be cool:

```
from .equations import residual_helmholtz_equation
from scipy import interpolate

residual_helmholtz_dt_dd = interpolate(method="spline", residual_helmholtz_equation.diff(dt, dd).evaluate, ...)
```